### DISCUSSION TITLE
Openapi-fetch with promise-like interface

### DISCUSSION THREAD

**Discussion by andrius-kurtinaitis** [openapi-ts/openapi-typescript#1316] [Announcements]
**Description**

When calling REST client functions, I expect to get Resolved with data when it was a success, and Reject with an error in case of failure.

**Proposal**

I'm trying to wrap for openapi-fetch like this:
```
export const { get, put, post, del } = createClient<paths>({ baseUrl: '/api' });

export const putClient = (url: Parameters<typeof put>[0], init: Parameters<typeof put>[1]) =>
  put(url, init)
    .catch((error: Error) => // network error - no response
      Promise.reject(errorToString(error)))
    .then(({ data, error, response }) =>
      data ?? Promise.reject(problemToString(response, error)));

```
But I can't get the types right. Apparently `url: Parameters<typeof put>[0], init: Parameters<typeof put>[1]` does not do the trick and wrong operations are getting picked from the generated ts file.
Any suggestions?

**Comment by drwpow** [comment]
You could maybe try the built-in `Parameters<T>` helper utility:

```ts
export const putClient = (...params: Parameters<typeof put>) => …
```

But I can’t say whether or not that will infer correctly.

Either way I would **NOT** try and declare a return type on the function; I think you’ll have a bad time trying to recreate the inference there.

This library is tricky because the TS inference is quite complex and hard to recreate. But if you import some of the helper types, you may find what you need (⚠️ though beware—those aren’t a guaranteed API and may change at any moment)

**Comment by drwpow** [comment]
The types got a little bit of an improvement recently, and openapi-fetch exports all its types. You can now probably just copy [the function signature for `PUT`](https://github.com/drwpow/openapi-typescript/blob/main/packages/openapi-fetch/src/index.ts#L95) without even needing to consume `openapi-fetch` directly (if you only import the types, it won’t cost any client weight).

I’m going to convert this into a discussion as I think this is resolved, but we can keep this open for any additional ideas / input.
